<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Almut Lütge" />

<meta name="date" content="2020-04-15" />

<title>Overall simulated batch characteristics</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/journal.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */

.sourceCode .row {
  width: 100%;
}
.sourceCode {
  overflow-x: auto;
}
.code-folding-btn {
  margin-right: -30px;
}
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>


<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">Overall simulated batch characteristics</h1>
<h4 class="author">Almut Lütge</h4>
<h4 class="date">15 April 2020</h4>

</div>


<div id="batch-characteristics-in-simulated-single-cell-rnaseq-data" class="section level2">
<h2>Batch Characteristics in simulated single cell RNAseq data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">suppressPackageStartupMessages</span>({
  <span class="kw">library</span>(scater)
  <span class="kw">library</span>(CellMixS)
  <span class="kw">library</span>(scales)
  <span class="kw">library</span>(purrr)
  <span class="kw">library</span>(circlize)
  <span class="kw">library</span>(viridis)
  <span class="kw">library</span>(jcolors)
  <span class="kw">library</span>(tidyr)
  <span class="kw">library</span>(dplyr)
  <span class="kw">library</span>(ComplexHeatmap)
  <span class="kw">library</span>(gridExtra)
  <span class="kw">library</span>(scran)
  <span class="kw">library</span>(magrittr)
  <span class="kw">library</span>(cowplot)
  <span class="kw">library</span>(colorspace)
})</code></pre></div>
<p>Datasets</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sce_name &lt;-<span class="st"> </span><span class="kw">unlist</span>(<span class="kw">c</span>(<span class="kw">strsplit</span>(params$sce_name, <span class="st">&quot;,&quot;</span>)))

sce_vec &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(params$sce, sce_name, <span class="st">&quot;_sce.rds&quot;</span>))
<span class="kw">names</span>(sce_vec) &lt;-<span class="st"> </span>sce_name

meta_vec &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(params$meta, sce_name, <span class="st">&quot;.rds&quot;</span>))
<span class="kw">names</span>(meta_vec) &lt;-<span class="st"> </span>sce_name

sce_small_vec &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(params$sce_norm, <span class="st">&quot;/&quot;</span>, sce_name, <span class="st">&quot;/sim_&quot;</span>, sce_name, <span class="st">&quot;_1__1_sce.rds&quot;</span>))
<span class="kw">names</span>(sce_small_vec) &lt;-<span class="st"> </span>sce_name
sum_vec &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">paste0</span>(params$summary, sce_name, <span class="st">&quot;.rds&quot;</span>))
<span class="kw">names</span>(sum_vec) &lt;-<span class="st"> </span>sce_name


cols &lt;-<span class="kw">c</span>(<span class="kw">c</span>(<span class="kw">jcolors</span>(<span class="st">&#39;pal6&#39;</span>),<span class="kw">jcolors</span>(<span class="st">&#39;pal8&#39;</span>))[<span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">8</span>,<span class="dv">14</span>,<span class="dv">5</span>,<span class="dv">2</span>:<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">7</span>,<span class="dv">9</span>:<span class="dv">13</span>,<span class="dv">15</span>:<span class="dv">20</span>)],<span class="kw">jcolors</span>(<span class="st">&#39;pal4&#39;</span>))
<span class="kw">names</span>(cols) &lt;-<span class="st"> </span><span class="kw">c</span>()</code></pre></div>
<div id="general" class="section level3">
<h3>General</h3>
<p>General summaries of the datasets as number of cells, genes, batches, batch origin</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">general_info &lt;-<span class="st"> </span><span class="kw">lapply</span>(sce_name, function(sce_nam){
  sce &lt;-<span class="st"> </span><span class="kw">readRDS</span>(sce_vec[sce_nam])
  meta &lt;-<span class="st"> </span><span class="kw">readRDS</span>(meta_vec[sce_nam])
  d &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;n_cells&quot;</span> =<span class="st"> </span><span class="kw">ncol</span>(sce),
                  <span class="st">&quot;n_genes&quot;</span> =<span class="st"> </span><span class="kw">nrow</span>(sce),
                  <span class="st">&quot;n_batches&quot;</span> =<span class="st"> </span><span class="kw">length</span>(<span class="kw">levels</span>(<span class="kw">as.factor</span>(<span class="kw">colData</span>(sce)[,meta$batch]))),
                  <span class="st">&quot;n_celltypes&quot;</span> =<span class="st"> </span><span class="kw">length</span>(<span class="kw">levels</span>(<span class="kw">as.factor</span>(<span class="kw">colData</span>(sce)[,meta$celltype]))),
                  <span class="st">&quot;linear&quot;</span> =<span class="st"> </span><span class="kw">mean</span>(sce$cms.Xadj1),
                  <span class="st">&quot;additive&quot;</span> =<span class="st"> </span><span class="kw">mean</span>(sce$cms.Xadj2),
                  <span class="st">&quot;interacting&quot;</span> =<span class="st"> </span><span class="kw">mean</span>(sce$cms.Xadj3),
                  <span class="st">&quot;reference&quot;</span> =<span class="st"> </span><span class="kw">mean</span>(sce$cms))
}) %&gt;%<span class="st"> </span><span class="kw">set_names</span>(sce_name) %&gt;%<span class="st"> </span><span class="kw">bind_rows</span>() %&gt;%<span class="st"> </span><span class="kw">set_rownames</span>(sce_name)

general_info$batch_origin &lt;-<span class="st"> </span><span class="kw">rep</span>(<span class="ot">NA</span>, <span class="kw">nrow</span>(general_info))
general_info[<span class="kw">c</span>(<span class="st">&quot;pbmc2_media&quot;</span>, <span class="st">&quot;pbmc_roche&quot;</span>, <span class="st">&quot;csf_media&quot;</span>), <span class="st">&quot;batch_origin&quot;</span>] &lt;-<span class="st"> &quot;media&quot;</span>
general_info[<span class="kw">c</span>(<span class="st">&quot;csf_patient&quot;</span>, <span class="st">&quot;kang&quot;</span>, <span class="st">&quot;pbmc2_pat&quot;</span>), <span class="st">&quot;batch_origin&quot;</span>] &lt;-<span class="st"> &quot;patient&quot;</span>
general_info[<span class="kw">c</span>(<span class="st">&quot;cellbench&quot;</span>, <span class="st">&quot;hca&quot;</span>), <span class="st">&quot;batch_origin&quot;</span>] &lt;-<span class="st"> &quot;protocol&quot;</span>
general_info$dataset &lt;-<span class="st"> </span><span class="kw">rownames</span>(general_info)

general_info &lt;-<span class="st"> </span>general_info %&gt;%<span class="st"> </span><span class="kw">arrange</span>(batch_origin)
<span class="kw">rownames</span>(general_info) &lt;-<span class="st"> </span>general_info$dataset</code></pre></div>
<p>Plot general infos</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#colors</span>
q_type &lt;-<span class="st"> </span><span class="kw">qualitative_hcl</span>(<span class="dv">4</span>, <span class="dt">palette =</span> <span class="st">&quot;Dark 3&quot;</span>)
col_type &lt;-<span class="st"> </span><span class="kw">list</span>(<span class="st">&quot;protocol&quot;</span> =<span class="st"> </span>q_type[<span class="dv">1</span>], 
                     <span class="st">&quot;media&quot;</span> =<span class="st"> </span>q_type[<span class="dv">3</span>], <span class="st">&quot;patient&quot;</span> =<span class="st"> </span>q_type[<span class="dv">4</span>])

col_type_vec &lt;-<span class="st"> </span><span class="kw">unlist</span>(col_type[general_info$batch_origin])

q_batch &lt;-<span class="st"> </span><span class="kw">sequential_hcl</span>(<span class="dv">5</span>, <span class="dt">palette =</span> <span class="st">&quot;BluYl&quot;</span>)
col_batch &lt;-<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">15</span>, <span class="dv">10</span>, <span class="dv">5</span>, <span class="dv">3</span>, <span class="dv">2</span>), q_batch)

<span class="co"># annotations</span>
h_cells &lt;-<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="dt">n_cells =</span> <span class="kw">anno_barplot</span>(general_info$n_cells, 
                                                    <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> col_type_vec),
                                                    <span class="dt">height =</span> <span class="kw">unit</span>(<span class="fl">2.2</span>, <span class="st">&quot;cm&quot;</span>)))

h_genes &lt;-<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="dt">n_genes =</span> <span class="kw">anno_barplot</span>(general_info$n_genes, 
                                                    <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> col_type_vec),
                                                    <span class="dt">height =</span> <span class="kw">unit</span>(<span class="fl">2.2</span>, <span class="st">&quot;cm&quot;</span>)))

<span class="co"># heatmap</span>
n_batches &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">as.matrix</span>(general_info[,<span class="kw">c</span>(<span class="st">&quot;n_batches&quot;</span>, <span class="st">&quot;n_celltypes&quot;</span>)])) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">set_colnames</span>(general_info$dataset)


h_batch &lt;-<span class="st"> </span><span class="kw">Heatmap</span>(n_batches, 
          <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
          <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">name =</span> <span class="st">&quot;celltype/batch&quot;</span>,
          <span class="dt">column_names_side =</span> <span class="st">&quot;bottom&quot;</span>,
          <span class="dt">column_names_rot =</span> <span class="dv">45</span>,
          <span class="dt">column_names_centered =</span> <span class="ot">TRUE</span>,
          <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
          <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, 
          <span class="dt">col =</span> col_batch,
          <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>),
          <span class="dt">show_heatmap_legend =</span> <span class="ot">FALSE</span>)



<span class="co">#legends</span>
lgd1 =<span class="st"> </span><span class="kw">Legend</span>(<span class="dt">labels =</span> <span class="kw">names</span>(col_type), <span class="dt">legend_gp =</span> <span class="kw">gpar</span>(<span class="dt">fill =</span> <span class="kw">unlist</span>(col_type)), 
              <span class="dt">title =</span> <span class="st">&quot;batch origin&quot;</span>)
lgd2 =<span class="st"> </span><span class="kw">Legend</span>(<span class="dt">col_fun =</span> col_batch, <span class="dt">at =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">15</span>), 
              <span class="dt">title =</span> <span class="st">&quot;celltype/batch&quot;</span>)
pd =<span class="st"> </span><span class="kw">packLegend</span>(lgd1, lgd2)

h_list &lt;-<span class="st">  </span>h_cells %v%<span class="st"> </span>h_genes %v%<span class="st"> </span>h_batch 

<span class="kw">draw</span>(h_list, <span class="dt">annotation_legend_list =</span> pd)</code></pre></div>
<p><img src="overall_sim_batch_characteristics_files/figure-html/plot%20general%20infos-1.png" width="672" /></p>
</div>
<div id="batch-sizestrength" class="section level3">
<h3>Batch size/strength</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">batch_size &lt;-<span class="st"> </span><span class="kw">lapply</span>(sce_name, function(sce_nam){
  sum &lt;-<span class="st"> </span><span class="kw">readRDS</span>(sum_vec[sce_nam])
  meta &lt;-<span class="st"> </span><span class="kw">readRDS</span>(meta_vec[sce_nam])
  d &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;mean_var_batch&quot;</span> =<span class="st"> </span>sum$mean_var_batch,
                  <span class="st">&quot;mean_var_celltype&quot;</span> =<span class="st"> </span>sum$mean_var_celltype,
                  <span class="st">&quot;mean_cms&quot;</span> =<span class="st"> </span>sum$mean_cms,
                  <span class="st">&quot;mean_de_genes&quot;</span> =<span class="st"> </span>sum$mean_mean_n_de_genes)
}) %&gt;%<span class="st"> </span><span class="kw">set_names</span>(sce_name) %&gt;%<span class="st"> </span><span class="kw">bind_rows</span>() %&gt;%<span class="st"> </span><span class="kw">set_rownames</span>(sce_name)

batch_size$batch_origin &lt;-<span class="st"> </span>general_info[<span class="kw">rownames</span>(batch_size), <span class="st">&quot;batch_origin&quot;</span>]
batch_size$dataset &lt;-<span class="st"> </span><span class="kw">rownames</span>(batch_size)
batch_size &lt;-<span class="st"> </span>batch_size %&gt;%<span class="st"> </span><span class="kw">arrange</span>(batch_origin)</code></pre></div>
<p>Plot batch size parameter</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#colors</span>
q_cms &lt;-<span class="st"> </span><span class="kw">sequential_hcl</span>(<span class="dv">5</span>, <span class="dt">palette =</span> <span class="st">&quot;YlOrRd&quot;</span>)
col_cms =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>), q_cms)

q_size &lt;-<span class="st"> </span><span class="kw">sequential_hcl</span>(<span class="dv">5</span>, <span class="dt">palette =</span> <span class="st">&quot;Reds&quot;</span>)
col_size &lt;-<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="fl">0.2</span>, <span class="fl">0.1</span>, <span class="fl">0.03</span>, <span class="fl">0.01</span>, <span class="dv">0</span>), q_size)

col_type &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;protocol&quot;</span> =<span class="st"> </span>q_type[<span class="dv">1</span>], <span class="st">&quot;media&quot;</span> =<span class="st"> </span>q_type[<span class="dv">3</span>], <span class="st">&quot;patient&quot;</span> =<span class="st"> </span>q_type[<span class="dv">4</span>])

q_size2 &lt;-<span class="st"> </span><span class="kw">sequential_hcl</span>(<span class="dv">5</span>, <span class="dt">palette =</span> <span class="st">&quot;Blues&quot;</span>)
col_size2 &lt;-<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="fl">0.65</span>, <span class="fl">0.5</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0.1</span>), q_size2)

<span class="co">#annotation</span>
ha_type =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="st">&quot;batch_origin&quot;</span> =<span class="st"> </span>batch_size$batch_origin,
                            <span class="dt">simple_anno_size =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;cm&quot;</span>),
                            <span class="dt">col =</span> <span class="kw">list</span>(<span class="st">&quot;batch_origin&quot;</span> =<span class="st"> </span>col_type), 
                            <span class="dt">annotation_height =</span> <span class="kw">unit</span>(<span class="fl">2.2</span>, <span class="st">&quot;cm&quot;</span>),
                            <span class="dt">annotation_name_side =</span> <span class="st">&quot;left&quot;</span>)

<span class="co">#heatmap</span>
<span class="co">#cms</span>
mean_cms &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">as.matrix</span>(batch_size$mean_cms))
<span class="kw">colnames</span>(mean_cms) &lt;-<span class="st"> </span>batch_size$dataset
<span class="kw">rownames</span>(mean_cms) &lt;-<span class="st"> &quot;mean_cms&quot;</span>

h_cms &lt;-<span class="st"> </span><span class="kw">Heatmap</span>(mean_cms, 
          <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
          <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">name =</span> <span class="st">&quot;cms&quot;</span>,
          <span class="dt">column_names_side =</span> <span class="st">&quot;bottom&quot;</span>,
          <span class="dt">column_names_rot =</span> <span class="dv">45</span>,
          <span class="dt">column_names_centered =</span> <span class="ot">TRUE</span>,
          <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
          <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, 
          <span class="dt">col =</span> col_cms,
          <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>))

<span class="co"># mean variance part</span>
n_batches &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">as.matrix</span>(batch_size[,<span class="kw">c</span>(<span class="st">&quot;mean_var_batch&quot;</span>, <span class="st">&quot;mean_var_celltype&quot;</span>)])) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">set_colnames</span>(batch_size$dataset)

h_size1 &lt;-<span class="st"> </span><span class="kw">Heatmap</span>(n_batches, 
          <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
          <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">name =</span> <span class="st">&quot;variance&quot;</span>,
          <span class="dt">column_names_side =</span> <span class="st">&quot;bottom&quot;</span>,
          <span class="dt">column_names_rot =</span> <span class="dv">0</span>,
          <span class="dt">column_names_centered =</span> <span class="ot">TRUE</span>,
          <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
          <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, 
          <span class="dt">col =</span> col_size,
          <span class="dt">top_annotation =</span> ha_type,
          <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>))


mean_de_genes &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">as.matrix</span>(batch_size$mean_de_genes))
<span class="kw">colnames</span>(mean_de_genes) &lt;-<span class="st"> </span>batch_size$dataset
<span class="kw">rownames</span>(mean_de_genes) &lt;-<span class="st"> &quot;mean_de_genes&quot;</span>

h_size2 &lt;-<span class="st"> </span><span class="kw">Heatmap</span>(mean_de_genes, 
          <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
          <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">name =</span> <span class="st">&quot;de genes [%]&quot;</span>,
          <span class="dt">column_names_side =</span> <span class="st">&quot;bottom&quot;</span>,
          <span class="dt">column_names_rot =</span> <span class="dv">0</span>,
          <span class="dt">column_names_centered =</span> <span class="ot">TRUE</span>,
          <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
          <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, 
          <span class="dt">col =</span> col_size2,
          <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>))

h_list &lt;-<span class="st">  </span>h_size1 %v%<span class="st"> </span>h_size2 %v%<span class="st"> </span>h_cms

<span class="kw">draw</span>(h_list)</code></pre></div>
<p><img src="overall_sim_batch_characteristics_files/figure-html/plot%20batch%20size-1.png" width="576" /></p>
</div>
<div id="batch-type" class="section level3">
<h3>Batch type</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#cols</span>
q_cms &lt;-<span class="st"> </span><span class="kw">sequential_hcl</span>(<span class="dv">5</span>, <span class="dt">palette =</span> <span class="st">&quot;Purple-Blue&quot;</span>)
col_cms =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.5</span>), q_cms)

<span class="co">#annotation</span>
ha_type =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="st">&quot;batch_origin&quot;</span> =<span class="st"> </span>batch_size$batch_origin, 
                            <span class="dt">simple_anno_size =</span>  <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;cm&quot;</span>),
                            <span class="dt">col =</span> <span class="kw">list</span>(<span class="st">&quot;batch_origin&quot;</span> =<span class="st"> </span>col_type), 
                            <span class="dt">annotation_height =</span> <span class="kw">unit</span>(<span class="fl">2.2</span>, <span class="st">&quot;cm&quot;</span>),
                            <span class="dt">annotation_name_side =</span> <span class="st">&quot;left&quot;</span>)

<span class="co">#heatmap</span>
<span class="co">#cms</span>
type &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">as.matrix</span>(general_info[,<span class="kw">c</span>(<span class="st">&quot;reference&quot;</span>,<span class="st">&quot;linear&quot;</span>, <span class="st">&quot;additive&quot;</span>, <span class="st">&quot;interacting&quot;</span>)])) %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">set_colnames</span>(batch_size$dataset)

h_cms &lt;-<span class="st"> </span><span class="kw">Heatmap</span>(type, 
          <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
          <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">name =</span> <span class="st">&quot;cms&quot;</span>,
          <span class="dt">column_names_side =</span> <span class="st">&quot;bottom&quot;</span>,
          <span class="dt">column_names_rot =</span> <span class="dv">45</span>,
          <span class="dt">column_names_centered =</span> <span class="ot">TRUE</span>,
          <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
          <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, 
          <span class="dt">col =</span> col_cms,
          <span class="dt">top_annotation =</span> ha_type,
          <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>))

<span class="kw">draw</span>(h_cms)</code></pre></div>
<p><img src="overall_sim_batch_characteristics_files/figure-html/batch%20type-1.png" width="672" /></p>
</div>
<div id="celltype-specificity" class="section level3">
<h3>Celltype specificity</h3>
<div id="variables" class="section level4">
<h4>Variables</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ct_spec &lt;-<span class="st"> </span><span class="kw">lapply</span>(sce_name, function(sce_nam){
  sum &lt;-<span class="st"> </span><span class="kw">readRDS</span>(sum_vec[sce_nam])
  meta &lt;-<span class="st"> </span><span class="kw">readRDS</span>(meta_vec[sce_nam])
  d &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;celltype_var_cms&quot;</span> =<span class="st"> </span>sum$celltype_var_cms,
                  <span class="st">&quot;mean_lfc_cor&quot;</span> =<span class="st"> </span>sum$mean_lfc_cor,
                  <span class="st">&quot;mean_de_overlap&quot;</span> =<span class="st"> </span>sum$mean_de_overlap)
}) %&gt;%<span class="st"> </span><span class="kw">set_names</span>(sce_name) %&gt;%<span class="st"> </span><span class="kw">bind_rows</span>() %&gt;%<span class="st"> </span><span class="kw">set_rownames</span>(sce_name)

ct_spec$batch_origin &lt;-<span class="st"> </span>general_info[<span class="kw">rownames</span>(ct_spec), <span class="st">&quot;batch_origin&quot;</span>]
ct_spec$dataset &lt;-<span class="st"> </span><span class="kw">rownames</span>(ct_spec)
ct_spec &lt;-<span class="st"> </span>ct_spec %&gt;%<span class="st"> </span><span class="kw">arrange</span>(batch_origin)</code></pre></div>
<p>Plot celltype specificity</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co">#colors</span>
q_cms &lt;-<span class="st"> </span><span class="kw">sequential_hcl</span>(<span class="dv">5</span>, <span class="dt">palette =</span> <span class="st">&quot;BluYl&quot;</span>)
col_cms =<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.002</span>, <span class="fl">0.004</span>, <span class="fl">0.006</span>, <span class="fl">0.008</span>), q_cms)

q_cor &lt;-<span class="st"> </span><span class="kw">rev</span>(<span class="kw">sequential_hcl</span>(<span class="dv">6</span>, <span class="dt">palette =</span> <span class="st">&quot;Reds&quot;</span>))
col_cor &lt;-<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.2</span>, <span class="fl">0.4</span>, <span class="fl">0.6</span>, <span class="fl">0.8</span>, <span class="fl">0.9</span>), q_cor)

col_type &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;protocol&quot;</span> =<span class="st"> </span>q_type[<span class="dv">1</span>], <span class="st">&quot;media&quot;</span> =<span class="st"> </span>q_type[<span class="dv">3</span>], <span class="st">&quot;patient&quot;</span> =<span class="st"> </span>q_type[<span class="dv">4</span>])

q_over &lt;-<span class="st"> </span><span class="kw">rev</span>(<span class="kw">sequential_hcl</span>(<span class="dv">6</span>, <span class="dt">palette =</span> <span class="st">&quot;Blues&quot;</span>))
col_over &lt;-<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="fl">0.1</span>, <span class="fl">0.2</span>, <span class="fl">0.3</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>), q_over)

<span class="co">#annotation</span>
ha_type =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="st">&quot;batch_origin&quot;</span> =<span class="st"> </span>ct_spec$batch_origin,
                            <span class="dt">simple_anno_size =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;cm&quot;</span>),
                            <span class="dt">col =</span> <span class="kw">list</span>(<span class="st">&quot;batch_origin&quot;</span> =<span class="st"> </span>col_type), 
                            <span class="dt">annotation_height =</span> <span class="kw">unit</span>(<span class="fl">2.2</span>, <span class="st">&quot;cm&quot;</span>),
                            <span class="dt">annotation_name_side =</span> <span class="st">&quot;left&quot;</span>)

<span class="co">#heatmap</span>
<span class="co">#cms</span>
var_cms &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">as.matrix</span>(ct_spec$celltype_var_cms))
<span class="kw">colnames</span>(var_cms) &lt;-<span class="st"> </span>ct_spec$dataset
<span class="kw">rownames</span>(var_cms) &lt;-<span class="st"> &quot;var_cms&quot;</span>

h_cms &lt;-<span class="st"> </span><span class="kw">Heatmap</span>(var_cms, 
          <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
          <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">name =</span> <span class="st">&quot;cms variance&quot;</span>,
          <span class="dt">column_names_side =</span> <span class="st">&quot;bottom&quot;</span>,
          <span class="dt">column_names_rot =</span> <span class="dv">45</span>,
          <span class="dt">column_names_centered =</span> <span class="ot">TRUE</span>,
          <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
          <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, 
          <span class="dt">col =</span> col_cms,
          <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>))

<span class="co"># correlation</span>
cor_tab &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">as.matrix</span>(ct_spec$mean_lfc_cor))
<span class="kw">colnames</span>(cor_tab) &lt;-<span class="st"> </span>ct_spec$dataset
<span class="kw">rownames</span>(cor_tab) &lt;-<span class="st"> &quot;lfc_correlation&quot;</span>

h_cor &lt;-<span class="st"> </span><span class="kw">Heatmap</span>(cor_tab, 
          <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
          <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">name =</span> <span class="st">&quot;lfc correlation&quot;</span>,
          <span class="dt">column_names_side =</span> <span class="st">&quot;bottom&quot;</span>,
          <span class="dt">column_names_rot =</span> <span class="dv">0</span>,
          <span class="dt">column_names_centered =</span> <span class="ot">TRUE</span>,
          <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
          <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, 
          <span class="dt">col =</span> col_cor,
          <span class="dt">top_annotation =</span> ha_type,
          <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>))


over &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">as.matrix</span>(ct_spec$mean_de_overlap))
<span class="kw">colnames</span>(over) &lt;-<span class="st"> </span>ct_spec$dataset
<span class="kw">rownames</span>(over) &lt;-<span class="st"> &quot;de_overlap&quot;</span>

h_over &lt;-<span class="st"> </span><span class="kw">Heatmap</span>(over, 
          <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, 
          <span class="dt">cluster_columns =</span> <span class="ot">FALSE</span>,
          <span class="dt">name =</span> <span class="st">&quot;de overlap [%]&quot;</span>,
          <span class="dt">column_names_side =</span> <span class="st">&quot;bottom&quot;</span>,
          <span class="dt">column_names_rot =</span> <span class="dv">0</span>,
          <span class="dt">column_names_centered =</span> <span class="ot">TRUE</span>,
          <span class="dt">column_names_gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">8</span>),
          <span class="dt">row_names_side =</span> <span class="st">&quot;left&quot;</span>, 
          <span class="dt">col =</span> col_over,
          <span class="dt">rect_gp =</span> <span class="kw">gpar</span>(<span class="dt">col=</span> <span class="st">&quot;white&quot;</span>))

h_list &lt;-<span class="st">  </span>h_cor %v%<span class="st"> </span>h_over %v%<span class="st"> </span>h_cms

<span class="kw">draw</span>(h_list)</code></pre></div>
<p><img src="overall_sim_batch_characteristics_files/figure-html/plot%20ct%20spec-1.png" width="672" /></p>
</div>
<div id="distributions" class="section level4">
<h4>Distributions</h4>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gene_info &lt;-<span class="st"> </span><span class="kw">lapply</span>(sce_name, function(sce_nam){
  <span class="kw">gc</span>()
  sce &lt;-<span class="st"> </span><span class="kw">readRDS</span>(sce_vec[sce_nam])
  meta &lt;-<span class="st"> </span><span class="kw">readRDS</span>(meta_vec[sce_nam])
  d &lt;-<span class="st"> </span><span class="kw">rowData</span>(sce)[, <span class="kw">grep</span>(<span class="st">&quot;logFC&quot;</span>, <span class="kw">names</span>(<span class="kw">rowData</span>(sce)))]
}) %&gt;%<span class="st"> </span><span class="kw">set_names</span>(sce_name)


ct_info &lt;-<span class="st"> </span><span class="kw">lapply</span>(sce_name, function(sce_nam){
  sce &lt;-<span class="st"> </span><span class="kw">readRDS</span>(sce_small_vec[sce_nam])
  meta &lt;-<span class="st"> </span><span class="kw">readRDS</span>(meta_vec[sce_nam])
  ct &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">colData</span>(sce)[, meta$celltype])
}) %&gt;%<span class="st"> </span><span class="kw">set_names</span>(sce_name)</code></pre></div>
</div>
</div>
<div id="logfc-distributions" class="section level3 tabset">
<h3>logFC distributions</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">plot_den &lt;-<span class="st"> </span>function(nam){
  mat &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(gene_info[[nam]])
  ct &lt;-<span class="st"> </span>ct_info[[nam]]
  <span class="co">#col and factors</span>
  ba_tab &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;batch&quot;</span> =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;_.*&quot;</span>,<span class="st">&quot;&quot;</span>,<span class="kw">colnames</span>(mat)),
                     <span class="st">&quot;celltype&quot;</span> =<span class="st"> </span><span class="kw">gsub</span>(<span class="st">&quot;.*_logFC_&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="kw">colnames</span>(mat)))
  ba_tab$cells &lt;-<span class="st"> </span>ct[ba_tab$celltype]
  col_den_b &lt;-<span class="st"> </span><span class="kw">rev</span>(cols)[<span class="dv">3</span>:(<span class="dv">2</span> +<span class="st"> </span><span class="kw">length</span>(<span class="kw">levels</span>(<span class="kw">as.factor</span>(ba_tab$batch))))] %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">set_names</span>(<span class="kw">levels</span>(<span class="kw">as.factor</span>(ba_tab$batch)))
  col_den_c &lt;-<span class="st"> </span>cols[<span class="dv">1</span>:<span class="kw">length</span>(<span class="kw">levels</span>(<span class="kw">as.factor</span>(ba_tab$celltype)))] %&gt;%<span class="st"> </span>
<span class="st">  </span><span class="kw">set_names</span>(<span class="kw">levels</span>(<span class="kw">as.factor</span>(ba_tab$celltype)))
  q_cells &lt;-<span class="st"> </span><span class="kw">sequential_hcl</span>(<span class="dv">3</span>, <span class="dt">palette =</span> <span class="st">&quot;Red-Blue&quot;</span>)
  col_cells &lt;-<span class="st"> </span><span class="kw">colorRamp2</span>(<span class="kw">c</span>(<span class="kw">max</span>(ct), <span class="kw">mean</span>(ct), <span class="kw">min</span>(ct)), q_cells)
  <span class="co">#annotations</span>
  ha_den =<span class="st"> </span><span class="kw">HeatmapAnnotation</span>(<span class="st">&quot;batch&quot;</span> =<span class="st"> </span>ba_tab$batch,
                            <span class="st">&quot;celltype&quot;</span> =<span class="st"> </span>ba_tab$celltype,
                            <span class="st">&quot;cells&quot;</span> =<span class="st"> </span>ba_tab$cells,
                            <span class="dt">simple_anno_size =</span> <span class="kw">unit</span>(<span class="dv">1</span>, <span class="st">&quot;cm&quot;</span>),
                            <span class="dt">col =</span> <span class="kw">list</span>(<span class="st">&quot;batch&quot;</span> =<span class="st"> </span>col_den_b,
                                       <span class="st">&quot;celltype&quot;</span> =<span class="st"> </span>col_den_c,
                                       <span class="st">&quot;cells&quot;</span> =<span class="st"> </span>col_cells),
                            <span class="dt">annotation_name_side =</span> <span class="st">&quot;left&quot;</span>)
  <span class="co">#heatmap</span>
  m &lt;-<span class="st"> </span><span class="kw">densityHeatmap</span>(mat, 
                    <span class="dt">ylim =</span> <span class="kw">c</span>(-<span class="dv">2</span>, <span class="dv">2</span>),
                    <span class="dt">title =</span> <span class="kw">paste0</span>(<span class="st">&quot;Density heatmap of logFC by celltype: &quot;</span>, nam),
                    <span class="dt">top_annotation =</span> ha_den,
                    <span class="dt">show_column_names =</span> <span class="ot">FALSE</span>
                    )
  <span class="kw">draw</span>(m)
}


template_density &lt;-<span class="st"> </span><span class="kw">c</span>(
    <span class="st">&quot;#### {{nam}}</span><span class="ch">\n</span><span class="st">&quot;</span>,
    <span class="st">&quot;```{r density{{nam}}, echo = FALSE}</span><span class="ch">\n</span><span class="st">&quot;</span>,
    <span class="st">&quot;m &lt;- plot_den(&#39;{{nam}}&#39;)&quot;</span>,
    <span class="st">&quot;```</span><span class="ch">\n</span><span class="st">&quot;</span>,
    <span class="st">&quot;</span><span class="ch">\n</span><span class="st">&quot;</span>
  )

plots_dens&lt;-<span class="st"> </span><span class="kw">lapply</span>(sce_name, 
  function(nam) knitr::<span class="kw">knit_expand</span>(<span class="dt">text =</span> template_density)
)</code></pre></div>
<div id="pbmc2_media" class="section level4">
<h4>pbmc2_media</h4>
<p><img src="overall_sim_batch_characteristics_files/figure-html/densitypbmc2_media-1.png" width="672" /></p>
</div>
<div id="csf_patient" class="section level4">
<h4>csf_patient</h4>
<p><img src="overall_sim_batch_characteristics_files/figure-html/densitycsf_patient-1.png" width="672" /></p>
</div>
<div id="csf_media" class="section level4">
<h4>csf_media</h4>
<p><img src="overall_sim_batch_characteristics_files/figure-html/densitycsf_media-1.png" width="672" /></p>
</div>
<div id="pbmc2_pat" class="section level4">
<h4>pbmc2_pat</h4>
<p><img src="overall_sim_batch_characteristics_files/figure-html/densitypbmc2_pat-1.png" width="672" /></p>
</div>
<div id="hca" class="section level4">
<h4>hca</h4>
<p><img src="overall_sim_batch_characteristics_files/figure-html/densityhca-1.png" width="672" /></p>
</div>
<div id="pbmc_roche" class="section level4">
<h4>pbmc_roche</h4>
<p><img src="overall_sim_batch_characteristics_files/figure-html/densitypbmc_roche-1.png" width="672" /></p>
</div>
<div id="kang" class="section level4">
<h4>kang</h4>
<p><img src="overall_sim_batch_characteristics_files/figure-html/densitykang-1.png" width="672" /></p>
</div>
<div id="cellbench" class="section level4">
<h4>cellbench</h4>
<p><img src="overall_sim_batch_characteristics_files/figure-html/densitycellbench-1.png" width="672" /></p>
</div>
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->
<script>
$(document).ready(function () {
  window.initializeCodeFolding("hide" === "show");
});
</script>

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
